### PR TITLE
Introduce Eslint

### PR DESCRIPTION
**Pull request by drets**
Not sure you want to restrict yourself by `eslint`, but I find `eslint` to be beneficial.

I also added `package-lock.json` (I think it's a good thing to have one on a long run).

Also, I messed up with `bom.ts` formatting. :smiley:

### CODE REVIEW

**Comment by drets** [comment]
> Cannot read property 'match' of undefined

Hehe

**Comment by drets** [comment]
I removed the `package-lock.json` since I am getting the following error:

> Cannot read property 'match' of undefined

I think I found the reason (source: https://npm.community/t/npm-err-cannot-read-property-match-of-undefined/234/3):

> The root cause is more interesting and less obvious and is implies that there’s an entry in the package-lock.json that’s missing a version.

So, no `package-lock.json` for now :)

**Comment by Kingwl** [comment]
I'm not sure do you noticed that we are using prettier😁

**Comment by drets** [comment]
> I'm not sure do you noticed that we are using prettier

Good morning. 

Yes, I noticed and run prettier on these changes. Does your comment directed to "Also, I messed up with bom.ts formatting."? If yes, I meant that WebStorm replaced the line separator, I guess, you are on Windows and I am on NixOS, I hope you can excuse me that. :smiley:

**Comment by drets** [comment]
> I'm not sure do you noticed that we are using prettier

After last commit CI checks that. :smile_cat:

File: src/value.ts [hunk 1]
```diff
@@ -21,18 +21,18 @@ export interface ConstantValueNumber extends ConstantValueBase {
   value: number
 }
 
-export interface COnstantValueBoolean extends ConstantValueBase {
+export interface ConstantValueBoolean extends ConstantValueBase {
```
Comment (Kingwl) [inline src/value.ts]: Aaaaaaaaaa....

**Comment by KagamiChan** [comment]
As we have eslint config/prettier, we could add them into the eslintrc, as some editors provides "autofix before saving", and we could avoid adding an extra step for prettier

**Comment by KagamiChan** [comment]
And on the linter config, do you think we could align with TypeScript repo's?

**Comment by drets** [comment]
> Thank you for your work, but I prefer to have fewer checks.

No problem, the only thing which I really liked about this change is explicit return type for the function. I like to see complete function types like in Haskell. 

Another cool thing to explore with eslint is type-aware rules.

> As we have eslint config/plugin for prettier, we could add them into the eslintrc, as some editors provide "autofix before saving" functionality, and we could avoid adding an extra step for prettier

Thanks for suggestion, fixed in https://github.com/HearTao/ts-bc/pull/16/commits/854cf4a02c51262f5813e5ee6093f5fad9f39480 (eslint runs prettier).

> And on the linter config, do you think we could align with TypeScript repo's?

Potentially, I guess, we could, but I just started from recommended eslint rules and than disabled the rules which were not obvious to fix for me.

**Pull request closed without merging**